### PR TITLE
fix(dag): roll cost-usd + failed-task metrics into DAG aggregate

### PR DESCRIPTION
**Pull request by H31MDALLR**
## Summary

Surfaced by the 2026-05-10 dogfood: a workflow consumed 25k+ tokens and reported \`Cost: \$0.0000\`; the \`:dag-result :metrics\` came back \`{:tokens 0 :cost-usd 0.0 :duration-ms 0}\` despite 8 sub-tasks each carrying real per-task metrics. **Two independent propagation gaps stacked into one symptom.**

- **Gap 1 (the dominant one) — ok/err shape mismatch in \`aggregate-results\`.** \`(get-in % [:data :metrics ...])\` works for \`dag/ok\` but \`dag/err\` puts the caller-supplied data under \`[:error :data ...]\`. With every dogfood task failing, 100% of per-task metrics dropped at the rollup. Added a private \`result-metrics\` accessor that handles both shapes; the aggregator now sums tokens / cost / duration uniformly across ok and err.
- **Gap 2 — \`:cost-usd\` dropped through the inner-loop shim.** \`dag-task-runner/create-generate-fn\` and \`create-repair-fn\` forwarded \`:tokens\` only, throwing away cost. The inner loop's \`update-metrics\` calls likewise threaded only :tokens + locally-computed :duration-ms. Both shims now forward :cost-usd; loop update-metrics calls (generate + repair) thread it through \`add-metric-values\`'s merge-with + so the value accumulates across iterations.

## Test plan

- [x] \`aggregate-results-sums-metrics-from-ok-and-err-test\` — one dag/ok (1000 tok, \$0.01) + one dag/err (2500 tok, \$0.03) ⇒ asserts rollup of (3500, 0.04, 17000). Pre-fix the err side contributed zero.
- [x] \`aggregate-results-handles-missing-metrics-gracefully-test\` — placeholder path with no \`:metrics\` keys returns zero rather than NPE.
- [x] Existing dag-orchestrator + dag-resilience-execution + loop inner + loop metrics-accumulation suites stay green.
- [x] Suite: 41 tests / 133 assertions across the affected namespaces; 0 failures / 0 errors. Full pre-commit suite green.

## What this doesn't fix

- **Cost calculation itself for the claude backend** when its CLI doesn't surface \`:total_cost_usd\`. Separate fix in the LLM client layer; this PR only ensures cost flows when upstream provides it.
- **Workflow-runner banner duration accounting** also wrong in the dogfood run (banner reported 9.7m vs ~50m wall). Different code path; separate fix.

Refs: \`~/.claude/projects/.../memory/project_dogfood_findings_2026_05_10.md\`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: components/workflow/test/ai/miniforge/workflow/dag_orchestrator_test.clj [hunk 1]
```diff
@@ -311,6 +311,45 @@
...
+            rollup the same way completed ones do."
+    (let [ok-result  (dag/ok  {:metrics {:tokens 1000 :cost-usd 0.01 :duration-ms 5000}})
+          err-result (dag/err :task-execution-failed
+                              "boom"
+                              {:metrics {:tokens 2500 :cost-usd 0.03 :duration-ms 12000}})
+          mixed      {:t-ok ok-result :t-err err-result}
+          agg        (dag-orch/aggregate-results mixed)]
+      (is (= 3500 (:total-tokens agg))
+          "ok + err token contributions both flow into the rollup")
+      (is (= 0.04 (:total-cost agg))
```
[the remark this answers is not in the document]
Comment (H31MDALLR) [reply components/workflow/test/ai/miniforge/workflow/dag_orchestrator_test.clj]: Fixed in 30f74bce — introduced cost-usd-epsilon (1.0e-6) and a small approx= helper. Cost assertion now uses (approx= expected actual) so 0.01 + 0.03 ≈ 0.039999... in IEEE doubles still passes. Tolerance is well below any rollup precision a caller relies on. Test fixture also lifted into named data (aggregate-rollup-fixture with :ok / :err / :sum keys) per the magic-number pass.

File: components/loop/src/ai/miniforge/loop/inner.clj [hunk 2]
```diff
@@ -384,7 +391,12 @@
                      result)
             loop-state (-> loop-state
                            (add-repair-attempt attempt)
+                           ;; Same :cost-usd forwarding as the
+                           ;; generate-step. Repair calls were
+                           ;; previously dropping cost on every
+                           ;; iteration.
                            (update-metrics {:tokens (:tokens-used result 0)
+                                            :cost-usd (:cost-usd result 0.0)
                                             :repair-calls 1}))]
```
[the remark this answers is not in the document]
Comment (H31MDALLR) [reply components/loop/src/ai/miniforge/loop/inner.clj]: Real bug. Verified against repair/attempt-repair — it returns one of make-{success,max-attempts,exhausted-strategies,escalation}-result, none of which set :tokens-used / :cost-usd at the top level. Per-attempt metrics live under :results. Fixed in 30f74bce: added private sum-repair-result-metrics that traverses :results and sums :tokens-used + :cost-usd into a {:tokens N :cost-usd N} map; loop/inner threads that into update-metrics. Two new tests pin the contract — aggregation across a 3-entry :results fixture (500 + 0 + 1200 tokens, 0.005 + 0.0 + 0.012 cost) plus an empty/missing-results path that returns zero-defaults instead of NPE.

File: components/workflow/src/ai/miniforge/workflow/dag_task_runner.clj [hunk 3]
```diff
@@ -56,15 +56,24 @@
...
     (let [result (agent/invoke impl-agent t context)]
+      ;; Forward the full metrics triple from the agent's response so
+      ;; the inner loop can roll cost + tokens up into :loop/metrics.
+      ;; Previously only :tokens was carried, leaving :cost-usd dropped
+      ;; on every DAG task and surfacing as $0.0000 in the runner
+      ;; banner despite real backend costs.
       {:artifact (:artifact result)
-       :tokens (or (get-in result [:metrics :tokens]) 0)})))
+       :tokens   (or (get-in result [:metrics :tokens]) 0)
+       :cost-usd (or (get-in result [:metrics :cost-usd]) 0.0)})))
```
[the remark this answers is not in the document]
Comment (H31MDALLR) [reply components/workflow/src/ai/miniforge/workflow/dag_task_runner.clj]: Fixed in 30f74bce — reworded the comment to reflect what the function actually does. It forwards :tokens + :cost-usd from the agent response; :duration-ms is measured loop-side from a wall-clock delta in inner.clj's generate step, not forwarded from here.
